tests/scripts/thread-cert/otci/otci/command_handlers.py

```diff
@@ -0,0 +1,199 @@
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+import logging
+import queue
+import re
+import threading
+import time
+from abc import abstractmethod
+from typing import Union, List
+
+from .connectors import OtCliHandler
+from .errors import ExpectLineTimeoutError, CommandError
+from .utils import match_line
+
+
+class OTCommandHandler:
+
+    @abstractmethod
+    def execute_command(self, cmd: str, timeout: float) -> List[str]:
+        pass
+
+    @abstractmethod
+    def close(self):
+        pass
+
+    @abstractmethod
+    def wait(self, duration: float) -> List[str]:
+        pass
+
+
+class OtCliCommandRunner(OTCommandHandler):
+    __PATTERN_COMMAND_DONE_OR_ERROR = re.compile(r'(Done|Error \d+:.*)$')
+
+    __PATTERN_LOG_LINE = re.compile(r'((\[(NONE|CRIT|WARN|NOTE|INFO|DEBG)\])'
+                                    r'|(-.*-+: )'  # e.g. -CLI-----: 
+                                    r')')
+    """regex used to filter logs"""
+
+    __ASYNC_COMMANDS = {
+        'scan',
+        'factoryreset',
+        'reset',
+    }
+
+    def __init__(self, otcli: OtCliHandler):
+        self.__otcli: OtCliHandler = otcli
+
+        self.__pending_lines = queue.Queue()
+        self.__should_close = threading.Event()
+        self.__otcli_reader = threading.Thread(target=self.__otcli_read_routine)
+        self.__otcli_reader.setDaemon(True)
+        self.__otcli_reader.start()
+
+    def __repr__(self):
+        return repr(self.__otcli)
+
+    def execute_command(self, cmd, timeout=10) -> None:
+        self.__otcli.writeline(cmd)
+
+        if cmd in {'reset', 'factoryreset'}:
+            return []
+
+        self.__expect_line(timeout, cmd)
+
+        output = self.__expect_line(timeout,
+                                    OtCliCommandRunner.__PATTERN_COMMAND_DONE_OR_ERROR,
+                                    asynchronous=cmd in OtCliCommandRunner.__ASYNC_COMMANDS)
+        return output
+
+    def wait(self, duration: float) -> List[str]:
+        self.__otcli.wait(duration)
+
+        output = []
+        try:
+            while True:
+                line = self.__pending_lines.get_nowait()
+                output.append(line)
+
+        except queue.Empty:
+            pass
+
+        return output
+
+    def close(self):
+        self.__should_close.set()
+        self.__otcli.close()
+
+    #
+    # Private methods
+    #
+
+    def __expect_line(self, timeout: float, expect_line: Union[str, re.Pattern], asynchronous=False) -> List[str]:
```

Comment:
Seems "re.Pattern" doesn't work in Python3.6. It works in Python3.8 but NDM has a pending issue with Python3.8.

Reply:
Yes, we are going to support Python3.6 and above. Will fix.